### PR TITLE
Support row comparisons.

### PR DESCRIPTION
**Pull request by petermattis**
See #1619. Still need to implement IN and NOT IN.

### CODE REVIEW

File: sql/parser2/eval.go [hunk 1]
```diff
@@ -108,8 +117,33 @@ func (d dstring) String() string {
 	return string(d)
 }
 
+type dtuple []Datum
+
+func (d dtuple) Bool() (dbool, error) {
+	return false, fmt.Errorf("cannot convert NULL to bool")
```
Comment (spencerkimball) [inline sql/parser2/eval.go]: This error message doesn't make sense.

File: sql/parser2/eval.go [hunk 2]
```diff
@@ -319,6 +352,10 @@ var cmpOps = map[cmpArgs]func(Datum, Datum) (Datum, error){
 	},
 }
 
+func init() {
```
Comment (tamird) [inline sql/parser2/eval.go]: why in init rather than the literal?

File: sql/parser2/eval.go [hunk 3]
```diff
@@ -639,3 +687,23 @@ func evalCaseExpr(expr *CaseExpr, env Env) (Datum, error) {
...
+func evalTupleEQ(ldatum, rdatum Datum) (Datum, error) {
+	left := ldatum.(dtuple)
+	right := rdatum.(dtuple)
+	if len(left) != len(right) {
+		return dbool(false), nil
+	}
+	for i, lval := range left {
+		d, err := evalComparisonOp(EQ, lval, right[i])
+		if err != nil {
+			return d, err
```
Comment (tamird) [inline sql/parser2/eval.go]: should this return `nil, err`?

File: sql/parser2/eval.go [hunk 4]
```diff
@@ -639,3 +687,23 @@ func evalCaseExpr(expr *CaseExpr, env Env) (Datum, error) {
...
 	return null, nil
 }
+
+func evalTupleEQ(ldatum, rdatum Datum) (Datum, error) {
+	left := ldatum.(dtuple)
+	right := rdatum.(dtuple)
+	if len(left) != len(right) {
+		return dbool(false), nil
+	}
+	for i, lval := range left {
```
Comment (tamird) [inline sql/parser2/eval.go]: nit: i'd use a normal for-loop here, but im not sure what the broadly accepted idiom here is

**Comment by tamird** [comment]
LGTM

File: sql/parser2/eval.go (hunk 1 above)
Comment (petermattis) [inline sql/parser2/eval.go]: Oops. Think I had a merge conflict with the multiple branches I've been working in.

File: sql/parser2/eval.go (hunk 2 above)
Comment (petermattis) [inline sql/parser2/eval.go]: This fixes an init-loop: `cmpOps -> evalTupleEq -> cmpOps`. I'll add a comment.

File: sql/parser2/eval.go (hunk 4 above)
Comment (petermattis) [inline sql/parser2/eval.go]: Stylistically a normal for-loop looks better. My understanding is that range loops which don't index into the slice being ranged avoid slice bounds checks. It definitely doesn't matter in this code. I'll change it.

File: sql/parser2/eval.go (hunk 3 above)
Comment (petermattis) [inline sql/parser2/eval.go]: `null, err`. Changed.

File: sql/parser2/eval.go (hunk 2 above)
Comment (tamird) [inline sql/parser2/eval.go]: :+1:

File: sql/parser2/eval.go [hunk 5]
```diff
@@ -395,10 +435,18 @@ func EvalExpr(expr Expr, env Env) (Datum, error) {
 		return null, fmt.Errorf("column \"%s\" not found", t)
 
 	case Tuple:
-		if len(t) != 1 {
-			return null, fmt.Errorf("unsupported expression type: %T: %s", expr, expr)
+		if len(t) == 1 {
+			return EvalExpr(t[0], env)
```
Comment (tschottdorf) [inline sql/parser2/eval.go]: Can this lead to unexpected behavior (something like `true == [true]`)? Why isn't this better off being treated as below?

**Comment by tschottdorf** [comment]
LGTM

**Pull request merged**